### PR TITLE
Added C wrappers for dense matrix differentiation

### PR DESCRIPTION
**Pull request by rdbyk**
Added `dense_matrix_{ diff, jacobian }` for elementwise differentiation and jacobian of a dense matrix.

### CODE REVIEW

**Comment by rdbyk** [comment]
Sorry, I know I should use clang-format, but I have currently only `clang-format-3.4` at hand which works not with `bin/test_format_local.sh` and for the time being I can't spot the problem using my eyes ... I will try to upgrade my machine within the next days and fix this.

**Comment by isuruf** [comment]
It's the difference between `CDenseMatrix* result` vs `CDenseMatrix *result`. Note the position of `*`. We use clang-format-3.7. So, using 3.4 is not going to work.
What OS are you using? clang-format-3.7 should be available in most package managers.

I've been meaning to create a web service that would push a commit automatically. Maybe it's time to start it.

**Comment by rdbyk** [comment]
Thanks, my eyes are really getting bad ;-)

Well, my machine is a bit outdated it has Ubuntu 14.04 LTS on it, and I am always a bit hesitant
regarding dist-upgrades ... when it is not really inevitable ...

**Comment by isuruf** [comment]
Ah, it's easy then. See the instructions at http://apt.llvm.org and replace 3.9 with 3.7
```
deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-3.7 main
deb-src http://apt.llvm.org/trusty/ llvm-toolchain-trusty-3.7 main
```

**Comment by rdbyk** [comment]
Thanks a lot, I will do that.

**Comment by rdbyk** [comment]
@isuruf  I have now finally installed clang-format-3.7 using
```
deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-3.7 main
deb-src http://apt.llvm.org/trusty/ llvm-toolchain-trusty-3.7 main
```
and it might be worth to remember this procedure:

```
wget -O - http://llvm.org/apt/llvm-snapshot.gpg.key|sudo apt-key add -
sudo apt-get update -y
sudo apt-get install clang-format-3.7
```

**Comment by rdbyk** [comment]
@isuruf it seems that one Travis CI build is hanging ... ??

File: symengine/cwrapper.cpp [hunk 1]
```diff
@@ -1148,6 +1148,25 @@ CWRAPPER_OUTPUT_TYPE dense_matrix_eye(CDenseMatrix *s, unsigned long int N,
     CWRAPPER_END
 }
 
+CWRAPPER_OUTPUT_TYPE dense_matrix_diff(CDenseMatrix *result,
+                                       const CDenseMatrix *A, basic const x)
+{
+    if (not is_a_Symbol(x))
+        return SYMENGINE_RUNTIME_ERROR;
```
Comment (ShikharJ) [inline symengine/cwrapper.cpp]: @rdbyk The `CodeCov` check is failing because this line is currently untested.

Comment (rdbyk) [inline symengine/cwrapper.cpp]: @ShikharJ `is_a_Symbol` gets tested several times, so how is one supposed to test a define like `SYMENGINE_RUNTIME_ERROR` ?

Comment (ShikharJ) [inline symengine/cwrapper.cpp]: @rdbyk I see. In that case, I guess this can be ignored.

**Comment by rdbyk** [comment]
@isuruf what do suggest to do, in order to remedy this code coverage thing?

**Comment by isuruf** [comment]
No. that's fine. code coverage is a guideline, not a must.

File: symengine/cwrapper.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/cwrapper.cpp]: A test would have been to have an integer (or anything other than a symbol) as input and check that the function returns a `SYMENGINE_RUNTIME_ERROR`.

**Pull request merged**